### PR TITLE
Introduce NodeJS codegen for Google Cloud Functions.

### PR DESCRIPTION
**Pull request by jmuk**
### PR checklist

- [x] Read the [contribution guildelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell/batch script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates)
- [x] Filed the PR against the correct branch: master for non-breaking changes and `2.3.0` branch for breaking (non-backward compatible) changes.

### Description of the PR

This PR introduces a new language "nodejs-google-cloud-functions." This is quite similar to nodejs server but the index.js is formed to work within [Google Cloud Functions]( https://cloud.google.com/functions/docs/).

### CODE REVIEW

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSGoogleCloudFunctionsCodegen.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        //   "controller.js")
+        // );
+        supportingFiles.add(new SupportingFile("swagger.mustache",
+                        "api",
+                        "swagger.yaml")
+        );
+        writeOptional(outputFolder, new SupportingFile("index.mustache", "", "index.js"));
+        writeOptional(outputFolder, new SupportingFile("package.mustache", "", "package.json"));
+        writeOptional(outputFolder, new SupportingFile("README.mustache", "", "README.md"));
+        if (System.getProperty("noservice") == null) {
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSGoogleCloudFunctionsCodegen.java]: @jmuk thanks for the PR. Instead of using `noservice` system property, what about using `java -Dmodels -DsupportingFiles` instead? https://github.com/swagger-api/swagger-codegen#selective-generation

You may also consider using `.swagger-codegen-ignore` to skip certain files when regenerating the code: https://github.com/swagger-api/swagger-codegen#ignore-file-format

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractNodeJSCodegen.java [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+public abstract class AbstractNodeJSCodegen extends DefaultCodegen implements CodegenConfig {
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractNodeJSCodegen.java]: @jmuk thanks for refactoring NodeJS generator into a common abstract class 👍

**Comment by wing328** [comment]
@jmuk thanks for the contribution.

> This PR introduces a new language "nodejs-google-cloud-functions." This is quite similar to nodejs server but the index.js is formed to work within Google Cloud Functions.

If only index.js is different, what about adding a CLI option to use different index.js mustache template instead of introducing another generator.

**Comment by jmuk** [comment]
To be more precise, not only `index.js` but also `package.json` (because it doesn't have to depend on some) and `README.md` are slightly different from nodejs-server -- but yeah, the contents are mostly common. Let me check how I can converge them into a single generator.

**Comment by wing328** [comment]
@jmuk One way is to pass `--additional-properties=enableGoogleCloudFunction=true` via command line argument and then you can use the following in the mustache template
```
{{#enableGoogleCloudFunction}}
... code for enabling Google cloud function here ...
{{/enableGoogleCloudFunction}}
```

**Comment by jmuk** [comment]
Updated. Now the single `NodeJSServerCodegen.java` does the codegen for GCF upon `googleCloudFunctions=true` additional property.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSGoogleCloudFunctionsCodegen.java (hunk 1 above)
Comment (jmuk) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSGoogleCloudFunctionsCodegen.java]: re: `.swagger-codegen-ignore`: removed

re: `noservice`: well, this is what currently nodejs-server is doing. I think it's better to do things like that in another PR.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java [hunk 3]
```diff
@@ -241,6 +249,46 @@ public String apiFileFolder() {
...
+                        "swagger.yaml")
+        );
+        if (getGoogleCloudFunctions()) {
+            writeOptional(outputFolder, new SupportingFile("index-gcf.mustache", "", "index.js"));
+        } else {
+            writeOptional(outputFolder, new SupportingFile("index.mustache", "", "index.js"));
+        }
+        writeOptional(outputFolder, new SupportingFile("package.mustache", "", "package.json"));
+        writeOptional(outputFolder, new SupportingFile("README.mustache", "", "README.md"));
+        if (System.getProperty("noservice") == null) {
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java]: Instead of using `noservice`, what about using selective generation as mentioned in https://github.com/swagger-api/swagger-codegen#selective-generation? Would that meet your need?

I prefer not to introduce another system property if the same output can already be achieved.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java [hunk 4]
```diff
@@ -76,27 +82,13 @@ public NodeJSServerCodegen() {
...
-        writeOptional(outputFolder, new SupportingFile("package.mustache", "", "package.json"));
-        writeOptional(outputFolder, new SupportingFile("README.mustache", "", "README.md"));
-        if (System.getProperty("noservice") == null) {
-            apiTemplateFiles.put(
-                    "service.mustache",   // the template to use
-                    "Service.js");       // the extension for each file to write
-        }
+        cliOptions.add(CliOption.newBoolean(GOOGLE_CLOUD_FUNCTIONS,
+            "When specified, it will generate the code which runs within Google Cloud Functions "
+                + "instead of standalone Node.JS server."));
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java]: SUGGESTION: A URL to tutorial on how to deploy the auto-generated code to Google Cloud Functions would be helpful for developers who are new to Google Cloud Functions.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java (hunk 3 above)
Comment (jmuk) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java]: Actually this comes from the original NodeJSCodegen.java, I just moved the code location. I'm not adding new.

Also, the `noservice` is for a different purpose from the selective generation.

The nodejs-server codegen generates two files per API. For the example of petstore, it generates files like `controllers/Pet.js` and `controllers/PetService.js` (see https://github.com/swagger-api/swagger-codegen/tree/master/samples/server/petstore/nodejs/controllers).

When this `noservice` is specified, it omits the generation of `PetService.js` file or any other -Service.js files. Only `Pet.js` or those files.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java (hunk 4 above)
Comment (jmuk) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java]: Great idea. Done.  Also added a section in README.mustache template, so people can refer to the URL after the code is generated.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java (hunk 3 above)
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/NodeJSServerCodegen.java]: @jmuk thanks for the clarification. I'll review and let you know if I've further question.

**Comment by wing328** [comment]
@jmuk PR merged into master. Thanks for your contribution.

**Comment by wing328** [comment]
FYI. I merged https://github.com/swagger-api/swagger-codegen/pull/4411 into master to improve the NodeJS code quality a bit.

File: modules/swagger-codegen/src/main/resources/nodejs-google-cloud-functions/index.mustache [hunk 5]
```diff
@@ -0,0 +1,44 @@
+'use strict';
```
Comment (wing328) [inline modules/swagger-codegen/src/main/resources/nodejs-google-cloud-functions/index.mustache]: @jmuk we'll move this file to nodejs/google-cloud-functions/index.mustache (and update the code accordingly). so that all nodejs (server) related files are under the folder.

Are you OK with the proposed change?

Comment (jmuk) [inline modules/swagger-codegen/src/main/resources/nodejs-google-cloud-functions/index.mustache]: Sure, I don't care the exact file location as long as the code works. But thank you for telling me!

Comment (wing328) [inline modules/swagger-codegen/src/main/resources/nodejs-google-cloud-functions/index.mustache]: I found that this file is the same as `index-gcf.mustache` so I simply removed it via https://github.com/swagger-api/swagger-codegen/pull/4808. I did some tests and the generator is still working properly with or without the option for Google cloud function.
